File: CS/GrowthWare.Framework.BusinessData/BusinessLogicLayer/BMessages.cs
using GrowthWare.Framework.BusinessData.DataAccessLayer.Interfaces;
using GrowthWare.Framework.Common;
using GrowthWare.Framework.Model.Profiles;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Data;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GrowthWare.Framework.BusinessData.BusinessLogicLayer
{
    /// <summary>
    /// Process business logic for messages
    /// </summary>
    /// <remarks>
    /// <![CDATA[
    /// MSecurityEntityProfile can be found in the GrowthWare.Framework.ModelObjects namespace.  
    /// 
    /// The following properties are necessary for correct business logic operation.
    /// .ConnectionString
    /// .DALName
    /// .DALNameSpace
    /// ]]>
    /// </remarks>
    /// <example> This sample shows how to create an instance of the class.
    /// <code language="VB.NET">BMessages
    /// <![CDATA[
    /// Dim myBll as new BMessages(mySecurityEntityProfile)
    /// ]]>
    /// </code>
    /// </example>
    public class BMessages : BaseBusinessLogic
    {
        private IDMessages m_DMessages;
        private MSecurityEntityProfile m_SecurityEntityProfile;

        /// <summary>
        /// Private BMessages() to ensure only new instances with passed parameters is used.
        /// </summary>
        /// <remarks></remarks>
        private BMessages()
        {
        }

        /// <summary>
        /// Parameters are need to pass along to the factory for correct connection to the desired data store.
        /// </summary>
        /// <param name="securityEntityProfile">The Security Entity profile used to obtain the DAL name, DAL name space, and the Connection String</param>
        /// <param name="centralManagement">Boolean value indicating if the system is being used to manage multiple database instances.</param>
        /// <remarks></remarks>
        /// <example> This sample shows how to create an instance of the class.
        /// <code language="VB.NET">
        /// <![CDATA[
        /// MSecurityEntityProfile mSecurityEntityProfile = MSecurityEntityProfile = New MSecurityEntityProfile();
        /// mSecurityEntityProfile.ID = ConfigSettings.DefaultSecurityEntityID;
        /// mSecurityEntityProfile.DAL = ConfigSettings.DAL;
        /// mSecurityEntityProfile.DAL_Namespace = ConfigSettings.DAL_NameSpace(mSecurityEntityProfile.DAL);
        /// mSecurityEntityProfile.DAL_Name = ConfigSettings.DAL_AssemblyName(mSecurityEntityProfile.DAL);
        /// mSecurityEntityProfile.ConnectionString = ConfigSettings.ConnectionString;
        /// 
        /// BMessages mBMessages = BMessages = New BMessages(mSecurityEntityProfile, ConfigSettings.CentralManagement);
        /// ]]>
        /// </code>
        /// <code language="C#">
        /// <![CDATA[
        /// Dim mSecurityEntityProfile As MSecurityEntityProfile = New MSecurityEntityProfile()
        /// mSecurityEntityProfile.ID = ConfigSettings.DefaultSecurityEntityID
        /// mSecurityEntityProfile.DAL = ConfigSettings.DAL
        /// mSecurityEntityProfile.DAL_Namespace = ConfigSettings.DAL_NameSpace(mSecurityEntityProfile.DAL)
        /// mSecurityEntityProfile.DAL_Name = ConfigSettings.DAL_AssemblyName(mSecurityEntityProfile.DAL)
        /// mSecurityEntityProfile.ConnectionString = ConfigSettings.ConnectionString
        /// 
        /// Dim mBMessages As BMessages = New BMessages(mSecurityEntityProfile, ConfigSettings.CentralManagement)
        /// ]]>
        /// </code>
        /// </example>
        public BMessages(MSecurityEntityProfile securityEntityProfile, bool centralManagement)
        {
            if (securityEntityProfile == null)
            {
                throw new ArgumentException("The securityEntityProfile and not be null!");
            }
            if (centralManagement)
            {
                if (m_DMessages == null)
                {
                    m_DMessages = (IDMessages)ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DMessages");
                }
            }
            else
            {
                m_DMessages = (IDMessages)ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DMessages");
            }
            m_SecurityEntityProfile = securityEntityProfile;
            m_DMessages.ConnectionString = securityEntityProfile.ConnectionString;
            //m_DMessages.SecurityEntitySeqID = SecurityEntityProfile.Id;
        }

        /// <summary>
        /// Gets all messages.
        /// </summary>
        /// <param name="securityEntitySeqId">The security entity ID.</param>
        /// <returns>Collection{MMessageProfile}.</returns>
        public Collection<MMessageProfile> GetMessages(int securityEntitySeqId)
        {
            Collection<MMessageProfile> mRetList = new Collection<MMessageProfile>();
            DataTable mDataTable = null;
            if (isDataBaseOnline()) 
            {
                try
                {
                    m_DMessages.Profile.SecurityEntitySeqId = securityEntitySeqId;
                    mDataTable = m_DMessages.Messages();
                    foreach (DataRow item in mDataTable.Rows)
                    {
                        mRetList.Add(new MMessageProfile(item));
                    }
                }
                catch (Exception)
                {
                    throw;
                }
                finally
                {
                    if (mDataTable != null)
                    {
                        mDataTable.Dispose();
                    }
                }            
            }
            return mRetList;
        }

        /// <summary>
        /// Purpose is to return data for a specific message.
        /// </summary>
        /// <param name="messageSeqId">int of the desired message profile object</param>
        /// <returns>DataRow</returns>
        /// <remarks></remarks>
        public DataRow GetMessage(int messageSeqId)
        {
            DataRow mRetVal = null;
            if (isDataBaseOnline()) mRetVal = m_DMessages.Message(messageSeqId);
            return mRetVal;
        }

        /// <summary>
        /// Adds the message.
        /// </summary>
        /// <param name="profile">The message profile.</param>
        /// <returns>System.Int32.</returns>
        public void Save(MMessageProfile profile)
        {
            if (profile == null) throw new ArgumentNullException("profile", "profile can not be null or Nothing in VB.Net!");
            if (isDataBaseOnline()) 
            {
                m_DMessages.Profile = profile;
                m_DMessages.Save();
            }
        }

        /// <summary>
        /// Searches the specified search criteria.
        /// </summary>
        /// <param name="searchCriteria">The search criteria.</param>
        /// <returns>DataTable.</returns>
        public DataTable Search(MSearchCriteria searchCriteria)
        {
            if (searchCriteria == null) throw new ArgumentNullException("searchCriteria", "searchCriteria can not be null (Nothing in VB) or empty!");
            DataTable mRetVal = null;
            if (string.IsNullOrEmpty(searchCriteria.WhereClause))
            {
                searchCriteria.WhereClause = " Security_Entity_SeqID = " + m_SecurityEntityProfile.Id.ToString(CultureInfo.InvariantCulture);
            }
            else
            {
                searchCriteria.WhereClause += " AND Security_Entity_SeqID = " + m_SecurityEntityProfile.Id.ToString(CultureInfo.InvariantCulture);
            }
            if (isDataBaseOnline()) mRetVal = m_DMessages.Search(searchCriteria);
            return mRetVal;
        }
    }
}

File: VB/GrowthWare.Framework.BusinessData/BusinessLogicLayer/BMessages.vb
Imports GrowthWare.Framework.Model.Profiles
Imports GrowthWare.Framework.BusinessData.DataAccessLayer.Interfaces
Imports GrowthWare.Framework.BusinessData.BusinessLogicLayer
Imports System.Collections.ObjectModel
Imports System.Globalization
Imports GrowthWare.Framework.Common

Namespace BusinessLogicLayer
    ''' <summary>
    ''' Process business logic for messages
    ''' </summary>
    ''' <remarks>
    ''' <![CDATA[
    ''' MSecurityEntityProfile can be found in the GrowthWare.Framework.ModelObjects namespace.  
    ''' 
    ''' The following properties are necessary for correct business logic operation.
    ''' .ConnctionString
    ''' .DALName
    ''' .DALNameSpace
    ''' ]]>
    ''' </remarks>
    ''' <example> This sample shows how to create an instance of the class.
    ''' <code language="VB.NET">
    ''' <![CDATA[
    ''' Dim myBll as new DMessages(mySecurityEntityProfile, ConfigSettings.CentralManagement)
    ''' ]]>
    ''' </code>
    ''' </example>
    Public Class BMessages
        Inherits BaseBusinessLogic

        Private m_DMessages As IDMessages
        Private m_SecurityEntityProfile As MSecurityEntityProfile

        ''' <summary>
        ''' Private sub New() to ensure only new instances with passed parameters is used.
        ''' </summary>
        ''' <remarks></remarks>
        Private Sub New()

        End Sub

        ''' <summary>
        ''' Parameters are need to pass along to the factory for correct connection to the desired datastore.
        ''' </summary>
        ''' <param name="SecurityEntityProfile">The Security Entity profile used to obtain the DAL name, DAL name space, and the Connection String</param>
        ''' <param name="CentralManagement">Boolean value indicating if the system is being used to manage multiple database instances.</param>
        ''' <remarks></remarks>
        ''' <example> This sample shows how to create an instance of the class.
        ''' <code language="VB.NET">
        ''' <![CDATA[
        ''' MSecurityEntityProfile mSecurityEntityProfile = MSecurityEntityProfile = New MSecurityEntityProfile();
        ''' mSecurityEntityProfile.ID = ConfigSettings.DefaultSecurityEntityID;
        ''' mSecurityEntityProfile.DAL = ConfigSettings.DAL;
        ''' mSecurityEntityProfile.DAL_Namespace = ConfigSettings.DAL_NameSpace(mSecurityEntityProfile.DAL);
        ''' mSecurityEntityProfile.DAL_Name = ConfigSettings.DAL_AssemblyName(mSecurityEntityProfile.DAL);
        ''' mSecurityEntityProfile.ConnectionString = ConfigSettings.ConnectionString;
        ''' 
        ''' BFunctions mBFunctions = BFunctions = New BFunctions(mSecurityEntityProfile, ConfigSettings.CentralManagement);
        ''' ]]>
        ''' </code>
        ''' <code language="C#">
        ''' <![CDATA[
        ''' Dim mSecurityEntityProfile As MSecurityEntityProfile = New MSecurityEntityProfile()
        ''' mSecurityEntityProfile.ID = ConfigSettings.DefaultSecurityEntityID
        ''' mSecurityEntityProfile.DAL = ConfigSettings.DAL
        ''' mSecurityEntityProfile.DAL_Namespace = ConfigSettings.DAL_NameSpace(mSecurityEntityProfile.DAL)
        ''' mSecurityEntityProfile.DAL_Name = ConfigSettings.DAL_AssemblyName(mSecurityEntityProfile.DAL)
        ''' mSecurityEntityProfile.ConnectionString = ConfigSettings.ConnectionString
        ''' 
        ''' Dim mBFunctions As BFunctions = New BFunctions(mSecurityEntityProfile, ConfigSettings.CentralManagement)
        ''' ]]>
        ''' </code>
        ''' </example>
        Public Sub New(ByVal securityEntityProfile As MSecurityEntityProfile, ByVal centralManagement As Boolean)
            If (securityEntityProfile Is Nothing) Then
                Throw New ArgumentException("securityEntityProfile can not be null or empty!")
            End If
            If Not centralManagement Then
                If m_DMessages Is Nothing Then
                    m_DMessages = ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DMessages")
                End If
            Else
                m_DMessages = ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DMessages")
            End If
            m_SecurityEntityProfile = securityEntityProfile
            m_DMessages.ConnectionString = securityEntityProfile.ConnectionString
            m_DMessages.SecurityEntitySeqId = securityEntityProfile.Id
        End Sub

        ''' <summary>
        ''' Gets the messages.
        ''' </summary>
        ''' <param name="securityEntitySeqId">The security entity seq ID.</param>
        ''' <returns>Collection{MMessageProfile}.</returns>
        Public Function GetMessages(ByVal securityEntitySeqId As Integer) As Collection(Of MMessageProfile)
            Dim mRetVal As Collection(Of MMessageProfile) = New Collection(Of MMessageProfile)
            Dim mDataTable As DataTable = Nothing
            If IsDatabaseOnline() Then
                Try
                    m_DMessages.SecurityEntitySeqId = securityEntitySeqId
                    mDataTable = m_DMessages.Messages()
                    For Each item As DataRow In mDataTable.Rows
                        Dim mProfile As MMessageProfile = New MMessageProfile(item)
                        mRetVal.Add(mProfile)
                    Next
                Catch ex As Exception
                    Throw
                Finally
                    If Not mDataTable Is Nothing Then
                        mDataTable.Dispose()
                    End If
                End Try
            End If
            Return mRetVal
        End Function

        ''' <summary>
        ''' Gets the messages.
        ''' </summary>
        ''' <param name="messageSeqId">The security entity seq ID.</param>
        ''' <returns>Collection{MMessageProfile}.</returns>
        Public Function GetMessage(ByVal messageSeqId As Integer) As MMessageProfile
            Dim mRetVal As MMessageProfile = Nothing
            If IsDatabaseOnline() Then
                mRetVal = New MMessageProfile(m_DMessages.GetMessage(messageSeqId))
            End If
            Return mRetVal
        End Function
        ''' <summary>
        ''' Save Function information to the database
        ''' </summary>
        ''' <param name="profile">MMessageProfile</param>
        Public Sub Save(ByVal profile As MMessageProfile)
            If profile Is Nothing Then Throw New ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!")
            m_DMessages.Profile = profile
            m_DMessages.Save()
        End Sub

        ''' <summary>
        ''' Returns a data table given the search criteria
        ''' </summary>
        ''' <param name="searchCriteria">MSearchCriteria</param>
        ''' <returns>DataTable</returns>
        ''' <remarks></remarks>
        Function Search(ByVal searchCriteria As MSearchCriteria) As DataTable
            If searchCriteria Is Nothing Then Throw New ArgumentNullException("searchCriteria", "searchCriteria cannot be a null reference (Nothing in Visual Basic)!")
            If String.IsNullOrEmpty(searchCriteria.WhereClause) Then
                searchCriteria.WhereClause = " Security_Entity_SeqID = " + m_SecurityEntityProfile.Id.ToString(CultureInfo.InvariantCulture)
            Else
                searchCriteria.WhereClause += " AND Security_Entity_SeqID = " + m_SecurityEntityProfile.Id.ToString(CultureInfo.InvariantCulture)
            End If
            Return m_DMessages.Search(searchCriteria)
        End Function
    End Class
End Namespace
